From: Andi Kleen <ak@suse.de>
Subject: [PATCH for 2.6.24rc2-rt1] Use real time pcp locking for page draining during cpu unplug


Use real time pcp locking for page draining during cpu unplug

Looks like a merging mistake that happened at some point. This
is the only place in the file that disables interrupts directly.

This fixes one case of CPU hotunplug failing on RT, but there 
are still more.

Signed-off-by: Andi Kleen <ak@suse.de>

---
 mm/page_alloc.c |    5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

Index: linux-2.6.24.4-rt4/mm/page_alloc.c
===================================================================
--- linux-2.6.24.4-rt4.orig/mm/page_alloc.c	2008-03-24 19:07:10.000000000 -0400
+++ linux-2.6.24.4-rt4/mm/page_alloc.c	2008-03-24 19:07:12.000000000 -0400
@@ -4058,10 +4058,11 @@ static int page_alloc_cpu_notify(struct 
 	int cpu = (unsigned long)hcpu;
 
 	if (action == CPU_DEAD || action == CPU_DEAD_FROZEN) {
-		local_irq_disable();
+		unsigned long flags;
+		__lock_cpu_pcp(&flags, cpu);
 		__drain_pages(cpu);
 		vm_events_fold_cpu(cpu);
-		local_irq_enable();
+		unlock_cpu_pcp(flags, cpu);
 		refresh_cpu_vm_stats(cpu);
 	}
 	return NOTIFY_OK;
